cirq/contrib/two_qubit_gates/math_utils.py

```diff
@@ -0,0 +1,292 @@
+                                           ) -> Tuple[np.ndarray, np.ndarray]:
+    """Generate random two-qubit unitaries.
+
+    Args:
+        num_samples: Number of samples.
+
+    Returns:
+        unitaries: tensor storing the unitaries, shape (num_samples,4,4).
+        kak_vecs: tensor storing KAK vectors of the unitaries (not canonical),
+            shape (num_samples, 3).
+
+    """
+    kl, kr = _two_local_2Q_unitaries(num_samples)
+
+    # Generate the non-local part by explict matrix exponentiation.
+    kak_vecs = np.random.rand(num_samples, 3) * np.pi  # / 4
+    # kak_vecs = np.sort(kak_vecs, axis=-1)[::-1]
+    A = KAK_vector_to_unitary(kak_vecs)
+    # Add a random phase
+    phases = np.random.rand(num_samples) * np.pi * 2
+    A = np.einsum('...,...ab->...ab', np.exp(1j * phases), A)
+
+    return np.einsum('...ab,...bc,...cd', kl, A, kr), kak_vecs
+
+
+def KAK_vector_to_unitary(vector: np.ndarray) -> np.ndarray:
+    r"""Convert a KAK vector to its unitary matrix equivalent.
+
+    Args:
+        vector: A KAK vector shape (..., 3). (Input may be vectorized).
+
+    Returns:
+        unitary: Corresponding 2-qubit unitary, of the form
+           exp( i k_x \sigma_x \sigma_x + i k_y \sigma_y \sigma_y
+                + i k_z \sigma_z \sigma_z).
+           matrix or tensor of matrices of shape (..., 4,4).
+    """
+    vector = np.asarray(vector)
+    gens = np.einsum('...a,abc->...bc', vector, _kak_gens)
+    evals, evecs = np.linalg.eigh(gens)
+
+    return np.einsum('...ab,...b,...cb', evecs, np.exp(1j * evals),
+                     evecs.conj())
+
+
+def _two_local_2Q_unitaries(num_samples):
+    kl_0 = random_qubit_unitary(num_samples)
+    kl_1 = random_qubit_unitary(num_samples)
+    kr_0 = random_qubit_unitary(num_samples)
+    kr_1 = random_qubit_unitary(num_samples)
+    kl = vector_kron(kl_0, kl_1)
+    kr = vector_kron(kr_0, kr_1)
+    return kl, kr
+
+
+def unitary_entanglement_fidelity(U_actual: np.ndarray,
+                                  U_ideal: np.ndarray) -> np.ndarray:
+    r"""Entanglement fidelity between two unitaries.
+
+    For unitary matrices, this is related to the average unitary fidelity F
+    as
+
+    :math:`F = \frac{F_e d + 1}{d + 1}`
+    where d is the matrix dimension.
+
+    Args:
+        U_actual : Matrix whose fidelity to U_ideal will be computed. This may
+            be a non-unitary matrix, i.e. the projection of a larger unitary
+            matrix into the computational subspace.
+        U_ideal : Unitary matrix to which U_actual will be compared.
+
+    Both arguments may be vectorized, in that their shapes may be of the form
+    (...,M,M) (as long as both shapes can be broadcast together).
+
+    Returns:
+        The entanglement fidelity between the two unitaries. For inputs with
+        shape (...,M,M), the output has shape (...).
+
+    """
+    U_actual = np.asarray(U_actual)
+    U_ideal = np.asarray(U_ideal)
+    if U_actual.shape[-1] != U_actual.shape[-2]:
```

Comment:
assert U_actual.shape[-1] != U_actual.shape[-2], "Inputs' trailing dimensions must be equal (square).")

Reply:
Done